pkg/game/board.go

```diff
@@ -172,11 +186,119 @@ func (board *board) isPositionValid(tile elements.PlacedTile) bool {
 	return true
 }
 
-//revive:disable-next-line:unused-parameter Until the TODO is finished.
 func (board *board) CanBePlaced(tile elements.PlacedTile) bool {
-	// TODO for future tasks:
-	// - implement generation of legal moves
-	// - to be implemented after #18, #19 and #20 to avoid code duplication
+	if !board.isPositionValid(tile) {
+		return false
+	}
+	if !slices.Contains(board.placeablePositions, tile.Position) {
+		return false
+	}
+
+	meepleCount := 0
+	featuresWithMeeples := map[feature.Type]struct{}{}
+	for _, feat := range tile.Features {
+		if feat.Meeple.Type != elements.NoneMeeple {
+			// Depending on use cases for this method, meeple counting may not be
+			// strictly necessary.
+			// One example where it's unnecessary is game.PlayTurn() which calls
+			// player.PlaceTile() and that already validates meeple count
+			// for other reasons.
+			meepleCount++
+			featuresWithMeeples[feat.FeatureType] = struct{}{}
+			if meepleCount > 1 {
+				return false
+			}
+		}
+	}
+
+	for feat := range featuresWithMeeples {
+		if !canBePlacedFunctions[feat](board, tile) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (board *board) cityCanBePlaced(tile elements.PlacedTile) bool {
+	return board.cityManager.CanBePlaced(tile)
+}
+
+func (board *board) fieldCanBePlaced(tile elements.PlacedTile) bool {
+	// While placing a tile, the player can only put a single meeple on a field.
+	// This means we don't have to care about whether our field expands into
+	// a different feature on our tile - we will only expand the feature
+	// if it has a meeple and that only happens once.
+	for _, feat := range tile.Features {
+		if feat.FeatureType != feature.Field || feat.Meeple.Type == elements.NoneMeeple {
+			continue
+		}
+
+		// unset the Meeple in our copy of the feature
+		feat.Meeple = elements.Meeple{}
+
+		// TODO: Field instance cannot expand to another feature on the tile checked by
+		// this method since it's not part of the board. See GH-86.
+		field := field.New(feat, tile.Position)
+		field.Expand(board, board.cityManager)
+
+		// score report function checks the whole field for placed meeples
+		// and reports any meeples that would be returned which we can use here
+		scoreReport := field.GetScoreReport()
+		if len(scoreReport.ReturnedMeeples) != 0 {
+			return false
+		}
+
+		// a single tile can only have one meeple on a field
+		break
+	}
+	return true
+}
+
+func (board *board) monasteryCanBePlaced(_ elements.PlacedTile) bool {
+	// meeple can always be placed on a monastery
+	return true
+}
+
+func (board *board) roadCanBePlaced(checkedTile elements.PlacedTile) bool {
+	for _, feat := range checkedTile.Features {
+		if feat.FeatureType != feature.Road || feat.Meeple.Type == elements.NoneMeeple {
+			continue
+		}
```

Comment:
Same thing as in fieldCanBePlaced

Reply:
See https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pull/62#discussion_r1756906486